convert2rhel/unit_tests/pkghandler_test.py

```diff
@@ -1320,21 +1320,115 @@ def test_fix_default_kernel_with_no_incorrect_kernel(self):
+        pytest.param("NetworkManager-1.18.8-2.0.1.el7_9", "1:NetworkManager-2.18.8-3.0.1.el7_9", -1, id="NVR&ENVR"),
+        pytest.param("2:NetworkManager-1.18.8-2.0.1.el7_9", "0:NetworkManager-1.18.8-3.0.1.el7_9", 1, id="ENVR"),
     ),
 )
 def test_compare_package_versions(version1, version2, expected):
     assert pkghandler.compare_package_versions(version1, version2) == expected
 
 
+@pytest.mark.parametrize(
+    ("version1", "version2", "exception_message"),
+    (
+        ("kernel-core-0:390-287.fc36", "kernel-0:390-287.fc36", "kernel-core does not match kernel"),
+        ("kernel-core-0:390-287.fc36.aarch64", "kernel-core-0:391-287.fc36.i86", "aarch64 does not match i86"),
+    ),
+)
+def test_compare_package_versions_warnings(version1, version2, exception_message):
+    with pytest.raises(ValueError, match=exception_message):
+        pkghandler.compare_package_versions(version1, version2)
+
+
+# format: name, epoch, version, release, arch
+PACKAGE_FORMATS = (
+    pytest.param(
+        "kernel-core-0:4.18.0-240.10.1.el8_3.i86", ["kernel-core", "0", "4.18.0", "240.10.1.el8_3", "i86"], id="NEVRA"
+    ),
+    pytest.param(
+        "kernel-core-0:4.18.0-240.10.1.el8_3", ["kernel-core", "0", "4.18.0", "240.10.1.el8_3", None], id="NEVR"
+    ),
+    pytest.param(
+        "1:NetworkManager-1.18.8-2.0.1.el7_9.aarch64",
+        ["NetworkManager", "1", "1.18.8", "2.0.1.el7_9", "aarch64"],
+        id="ENVRA",
+    ),
+    pytest.param(
+        "1:NetworkManager-1.18.8-2.0.1.el7_9", ["NetworkManager", "1", "1.18.8", "2.0.1.el7_9", None], id="ENVR"
+    ),
+    pytest.param(
+        "NetworkManager-1.18.8-2.0.1.el7_9.aarch64",
+        ["NetworkManager", None, "1.18.8", "2.0.1.el7_9", "aarch64"],
+        id="NVRA",
+    ),
+    pytest.param(
+        "NetworkManager-1.18.8-2.0.1.el7_9", ["NetworkManager", None, "1.18.8", "2.0.1.el7_9", None], id="NVR"
+    ),
+)
+
+
+@pytest.mark.skipif(pkgmanager.TYPE == "yum", reason="cannot test dnf backend if dnf is not present")
+@pytest.mark.parametrize(
+    ("package", "expected"),
+    (PACKAGE_FORMATS),
+)
+def test_parse_pkg_string_dnf(package, expected, monkeypatch):
+    assert pkghandler.parse_pkg_string(package) == expected
+    parse_pkg_with_dnf_mock = mock.Mock()
+    monkeypatch.setattr(pkghandler, "_parse_pkg_with_dnf", value=parse_pkg_with_dnf_mock)
+    pkghandler.parse_pkg_string(package)
+    parse_pkg_with_dnf_mock.assert_called_once()
+
+
+@pytest.mark.skipif(pkgmanager.TYPE == "dnf", reason="cannot test yum backend if yum is not present")
+@pytest.mark.parametrize(
+    ("package", "expected"),
+    (PACKAGE_FORMATS),
+)
+def test_parse_pkg_string_yum(package, expected, monkeypatch):
+
+    assert pkghandler.parse_pkg_string(package) == expected
+    parse_pkg_with_yum_mock = mock.Mock()
+    monkeypatch.setattr(pkghandler, "_parse_pkg_with_yum", value=parse_pkg_with_yum_mock)
+    pkghandler.parse_pkg_string(package)
+    parse_pkg_with_yum_mock.assert_called_once()
```

Comment:
I think you could unify those methods and have just one that asserts you're going through the correct path. 

They are not adding too much value the way they are right now compared to the `test_parse_pkg_with_yum` and `test_parse_pkg_with_dnf`

Some ideas that could be done: 

1. Remove the `test_parse_pkg_with_[dnf, yum]` and have only those two that will call them (without mocks) and assert the result
2. Have just one test that will assert the path inside the function. This test would be backend independent, you would mock this dynamically based on a parametrization.
3. You can leave the two methods there, but remove the first part where you call `assert pkghandler.parse_pkg_string(package) == expected` as you're calling it twice, but there is no need for this.

If you have more ideas, I would love to hear too.

Reply:
The third idea is what I had originally so I'll go back to that format, thanks!